scenarioColors.js

```diff
@@ -1,42 +1,62 @@
 const ScenarioSdk = require('@qasymphony/scenario-sdk');
 
-const StepSdk = {
-    getStepSdk(qtestToken, scenarioProjectId) {
-        return new ScenarioSdk.Steps({ qtestToken, scenarioProjectId });
-    }
+// @pure
+const state = {
+    sdk: null
 }
 
-const Steps = {
-    updateStepResults(stepSdk, name, status) {
-        return stepSdk.getSteps(`"${name}"`).
-            then(steps => Promise.all(steps.map(step => stepSdk.updateStep(step.id, Object.assign(step, { status })))))
-            .catch(function (err) {
-                console.log('Error updating colors: ' + err);
-            });
-    }
-};
-
-exports.handler = function ({ event: body, constants, triggers }, context, callback) {
-    var payload = body;
-    var testLogs = payload.logs;
-
-    stepSdk = StepSdk.getStepSdk(constants.QTEST_TOKEN, constants.SCENARIO_PROJECT_ID);
-    
-    for (var res of testLogs) {
-        for (var step of res["test_step_logs"]) {
-            var stepName = step.description;
-            var stepStatus = step.status;
-
-            // Undefined means no step definition existed and it should fail
-            if (stepStatus == "undefined") {
-                stepStatus = "failed";
-            }
-
-            // one of PASSED (green), FAILED (red), or SKIPPED (yellow)
-            stepStatus = stepStatus.toUpperCase();
-
-            // Call the pulse API to update step results
-            Steps.updateStepResults(stepSdk, stepName, stepStatus);
-        }
-    }
+// @impure: Data returned is dependent on where the ScenarioSDK is pointing
+const getStepSdk = (qtestToken, scenarioProjectId) => new ScenarioSdk.Steps({ qtestToken, scenarioProjectId });
+
+// @impure: has side effects (updates ScenarioSdk)
+const setConfigAndRetrieveSDK = (scenarioUrl, qtestToken, scenarioProjectId) => {
+    ScenarioSdk.config({ scenarioUrl: scenarioUrl });
+    return getStepSdk(qtestToken, scenarioProjectId);
 }
+
+// @impure: uses I/O 
+const loggingOutError = err => console.log('Error updating colors: ' + err);
+
+// @impure: reliant on state object
+const addingStatusToStep = (sdkStep, status) => state.sdk.updateStep(sdkStep.id, Object.assign(sdkStep, { "status": status }));
+
+// @impure: reliant on network
+const asyncronouslyUpdateTheStatusOfEachStep = (sdkSteps, status) => Promise.all(sdkSteps.map((sdkStep) => addingStatusToStep(sdkStep, status)));
+
+// Grabs all of the steps matching the step description using the Scenario API and then updates there status based on the test results
+// Note - Whilst the test results can feature a step multiple times, the API only appears to return a single instance, so the status will get set multiple times, and if there is 
+// different a statuses the last one set will win  E.g. in one instance the step passed and in another it failed you are at the mercy of the order in which the promises resolve as to
+// which gets set in Scenario.
+// @impure: reliant on state object
+const updateStepResults = (name, status) => {
```

Comment:
Could you use the `async` library or something similar to to enforce the order this is set in?

Reply:
More than likely.  In fact I'm sure there are several ways where we could actively prioritise setting one status over the other, but I'm not sure which would be more desirable.  E.g. If we have a step that ran twice and in one instance it passes and in one it fails.  If we took the passed result we would be masking a failure, but if we took a failed result the output of the scenario where the step passed would be confusing.  (It would appear as if it continued to run other steps after a step failed.)

This comment is more to highlight that there is a limitation within the Scenario plugin itself.  IMO we shouldn't be trying to map multiple statuses into a single status because it doesn't make sense.  I would expect a 1 to 1 relationship and I intend on raising this with Tricentis to see if it's something they could support.

Reply:
That makes sense, for me that'd be a more helpful comment in the code. Or at least the commit message :)

Reply:
I've updated the comment and given the change a meaningful commit message.